src/main/kotlin/com/terraformation/backend/accelerator/db/ParticipantProjectSpeciesStore.kt

```diff
@@ -89,6 +94,31 @@ class ParticipantProjectSpeciesStore(
         .execute()
   }
 
+  fun fetchLastUpdatedSpeciesTime(projectId: ProjectId): Instant {
+    requirePermissions { readProject(projectId) }
+
+    val user = currentUser()
+
+    val lastUpdatedTime =
+        with(PARTICIPANT_PROJECT_SPECIES) {
+          dslContext
+              .select(ID, MODIFIED_TIME)
+              .from(this)
+              .where(PROJECT_ID.eq(projectId))
+              .orderBy(MODIFIED_TIME.desc())
+              .limit(1)
+              .fetch()
+              .filter { user.canReadParticipantProjectSpecies(it[ID]!!) }
```

Comment:
Obviously in the current implementation this will always return true if the user has permission to read the project at all, but in theory, this filter will cause the method to have some unexpected behavior:

1. You call this method, get back a timestamp
2. Someone else modifies a participant project species you don't have permission to read
3. You call this method, it throws a "not found" exception even though there are still other species you can read

Three possible solutions here, I think:

1. Get rid of the per-species permission check entirely, and say that if you have permission to read the project, you have permission to ask when its most recent participant project species update was
2. Get rid of the `limit(1)` so it'll return the most recent timestamp that's visible to you (this will of course cause the query to fetch the full list of results but hopefully the list won't be massive)
3. Bake the permission filtering into the query

Those are in descending order of my personal preference but any of them would work.

Reply:
I didn't think about that, thanks. I will go with option 1.